### PR TITLE
Fix #6226: Added closestWidgetVar EL function

### LINKED ISSUE #6226: EL function closestWidgetVar
Currently you have an EL method `resolveWidgetVar(String expression, UIComponent source)` to find a widget var. I would like to have a method to find the widget var relative to the current component, so go up to the root until a component which implements `Widget` is found.

I created this method in my project:
```
  public static String closestWidgetVar() {
    return Stream.iterate(Components.getCurrentComponent().getParent(),
                          c -> Objects.nonNull(c.getParent()),
                          UIComponent::getParent)
            .filter(Widget.class::isInstance)
            .findFirst()
            .map(Widget.class::cast)
            .map(Widget::resolveWidgetVar)
            .map(w -> String.format("PF('%s')", w))
            .orElse(null);
  }
```
It uses Java 9 and OmniFaces, so it needs a bit of rewriting. If you want to I can create a PR for this.

### PR DESCRIPTION
**Pull request by jepsar**
Fixes #6226

### CODE REVIEW

File: src/main/java/org/primefaces/expression/SearchExpressionUtils.java [hunk 1]
```diff
@@ -85,10 +87,30 @@ public static String resolveWidgetVar(String expression, UIComponent component)
...
-            return "PF('" + ((Widget) resolvedComponent).resolveWidgetVar(context) + "')";
+            return javascriptWidgetVar((Widget) resolvedComponent, context);
         }
         else {
             throw new FacesException("Component with clientId " + resolvedComponent.getClientId() + " is not a Widget");
         }
     }
+
+    // used by p:closestWidgetVar
+    public static String closestWidgetVar() {
```
Comment (jepsar) [inline src/main/java/org/primefaces/expression/SearchExpressionUtils.java]: I wasn't sure where to put these methods... is it okay to have them here?

**Review by melloware — APPROVED** [review]
Looks good to me but will let Thomas review

**Comment by tandraschko** [comment]
i dont like your change that you automatically add PF('') now
this breaks backward compatibility and doesnt allow to just resolve the widgetVar anymore, without having PF('') around

**Comment by jepsar** [comment]
> i dont like your change that you automatically add PF('') now
> this breaks backward compatibility and doesnt allow to just resolve the widgetVar anymore, without having PF('') around

I didn't change that, I just moved it to a method to DRY things up and fixed the documentation.

https://github.com/primefaces/primefaces/blob/de9a16e875e66c7233aa30fb714b3be7409c2c02/src/main/java/org/primefaces/expression/SearchExpressionUtils.java#L88

.. but I agree it's a bit odd. I was surprised it was working like that. Shall I remove it?

**Comment by tandraschko** [comment]
oha, i see
please, it should just return the varName, not PF('')

**Comment by tandraschko** [comment]
and IMO we should not create 2 methods for it with different names, we had it in the past and its dirty

i think we have 2 solutions:
- test overloading with all current containers (it didnt work in the past)
or
- always pass the UIComponent like we did for resolveClientId. So you always need: #{p:closestWidgetVar(component)}

#{component} like like UIComponet.getCurrentComponent

**Comment by jepsar** [comment]
@tandraschko done

**Comment by tandraschko** [comment]
Please also adjust the migration guide that people must add PF('') now

**Comment by jepsar** [comment]
> Please also adjust the migration guide that people must add PF('') now

Done as well

**Pull request merged**